scripts/create-darwin-volume.sh

```diff
@@ -127,42 +164,75 @@ main() {
-                exit 1
-            fi
+    if [ -n "$create_volume" ]; then
+        echo "Creating a Nix volume..." >&2
+
+        sudo diskutil apfs addVolume "$disk" "$NIX_VOLUME_FS" "$volume" -mountpoint /nix
+        new_uuid="$(volume_uuid "$volume")"
+
+        if [ "$INSTALL_MODE" = "no-daemon" ]; then # exported by caller
+            # TODO: is there a better way to do this?
+            sudo chown $USER:admin /nix
         fi
 
-        sudo diskutil apfs addVolume "$disk" APFS 'Nix Store' -mountpoint /nix
-        volume="Nix Store"
+        if test_filevault_in_use; then
+            # security program's flags won't let us both specify a keychain
+            # and be prompted for a pw to add; two step workaround:
+            # 1. add a blank pw to system keychain
+
+            # system is in some sense less secure than user keychain... (it's
+            # possible to read the password for decrypting the keychain) but
+            # the user keychain appears to be available too late. As far as I
+            # can tell, the file with this password (/var/db/SystemKey) is
+            # inside the FileVault envelope. If that isn't true, it may make
+            # sense to store the password inside the envelope?
+            sudo /usr/bin/security add-generic-password -a "$volume" -s "$new_uuid" -D "$volume encryption password" -j "Added automatically by the Nix installer for use by /Library/LaunchDaemons/org.nixos.darwin-store.plist" "/Library/Keychains/System.keychain"
+            # TODO: decide if we should add `-T /System/Library/CoreServices/APFSUserAgent`
+            # This should let the system seamlessly supply the password for this volume
+            # which in turn means the fstab entry is enough for the system to (eventually)
+            # decrypt and mount the volume we're adding, but I hesitate because I'm not
+            # certain the system _should_ transparently failover if the LaunchDaemon is
+            # broken for some reason? Without supplying this flag, the system will instead
+            # start prompting them to allow APFSUserAgent to access this credential.
+
+            # 2. add a password with the -U (update) flag and -w (prompt if last)
+            #    flags, but specify no keychain; security will use the first it finds
+            prepare_darwin_volume_password "$volume" "$new_uuid" | sudo diskutil apfs encryptVolume "$volume" -user disk -stdinpassphrase
+        fi
     else
         echo "Using existing '$volume' volume" >&2
     fi
 
-    if ! test_fstab; then
-        echo "Configuring /etc/fstab..." >&2
-        label=$(echo "$volume" | sed 's/ /\\040/g')
-        # shellcheck disable=SC2209
-        printf "\$a\nLABEL=%s /nix apfs rw,nobrowse\n.\nwq\n" "$label" | EDITOR=ed sudo vifs
+    if ! test_voldaemon; then
+        echo "Configuring LaunchDaemon to mount '$volume'..." >&2
+        generate_mount_daemon | sudo tee /Library/LaunchDaemons/org.nixos.darwin-store.plist >/dev/null
```

Comment:
We need to check if the volume is encrypted first. If this is using a pre-existing unencrypted Nix volume, then the LaunchDaemon is superfluous (and the keychain entry won't exist).

For that matter, if it's using a pre-existing Nix volume at all, then there's no guarantee about the keychain entry's status. We should probably actually just skip this if we're not creating the volume. Anyone who creates their own volume is then on the hook for managing its mounting, and we can print a warning to this effect in the "Using existing volume" path.

Reply:
I've wrestled with this a bit. The latest update may resolve it, though that's contingent on what people think of the "curing" approach (and it likely needs logical refinement).

Basically, I've cut attempts to infer user intent or do anything magic based on what exists. If the root is read-only and `NIX_VOLUME_CREATE=1` (the default), the installer will want to create a volume (encrypted if filevault is on), and the "curing" version currently enabled asks the user for permission to remove anything that conflicts (volume, keychain credentials, fstab, synthetic.conf, daemon) before it'll install.

I've gone to a little pain to set up create-darwin-volume.sh so that it'll still behave roughly as it currently does if invoked with no arguments, but to also support passing a function to call when you source or invoke it. I won't say there's a clear path to it yet since I haven't thought through the logistics of scripting this exact case, but I'm inclined to close off some options for now and focus on getting the golden path right--but make the installer scripts flexible enough that people can leverage them rather than just copy and customize them?